### PR TITLE
paperless-ngx: 2.0.1 -> 2.1.1

### PR DESCRIPTION
**Pull request by ajaxbits**
## Description of changes

Bumps paperless-ngx to v2.1.1.

With the addition of the `canvas` js library to paperless-ngx [here](https://github.com/paperless-ngx/paperless-ngx/commit/826322b6105c45c103f0da8d482121ecb3ff50d2), we need to add `pkg-config` and `pango` to the buildInputs. Note that the package only compiles if `pango` is in `buildInputs`, not `nativeBuildInputs`.

I've tested this successfully on x86_linux, but there's an error compiling on my aarch64-darwin machine. This stems from errors in `mupdf`, propogating up to `img2pdf`. A PR is open to fix these issues currently: https://github.com/NixOS/nixpkgs/pull/272807#issuecomment-1847347779

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/office/paperless-ngx/default.nix [hunk 1]
```diff
@@ -19,13 +19,13 @@
...
 let
-  version = "2.0.1";
+  version = "2.1.0";
 
   src = fetchFromGitHub {
     owner = "paperless-ngx";
     repo = "paperless-ngx";
-    rev = "refs/tags/v${version}";
-    hash = "sha256-qSX+r99y7a/eITfaC/UYqSgcxx/xYOqJ4tY/iuvoeNA=";
+    rev = "v2.1.0";
```
Comment (SuperSandro2000) [inline pkgs/applications/office/paperless-ngx/default.nix:27]: Please revert the change in rev

**Comment by ajaxbits** [comment]
Result of `nixpkgs-review pr 272563` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>paperless-ngx</li>
  </ul>
</details>

**Comment by ajaxbits** [comment]
Coming back to this tomorrow 12/7 -- anyone can pick up the work in the meantime if they want

**Comment by ajaxbits** [comment]
Result of `nixpkgs-review pr 272563` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>paperless-ngx</li>
  </ul>
</details>

File: pkgs/applications/office/paperless-ngx/default.nix [hunk 2]
```diff
@@ -46,14 +54,26 @@ let
...
+      cairo
+      giflib
+      libjpeg
+      libpng
+      librsvg
+      pango
+      pkg-config
+    ] ++ lib.optionals stdenv.isDarwin [
+      mupdf
+    ];
```
Comment (SuperSandro2000) [inline pkgs/applications/office/paperless-ngx/default.nix:75]: This is probably because of https://github.com/paperless-ngx/paperless-ngx/commit/826322b6105c45c103f0da8d482121ecb3ff50d2 and the new canvas dependency

File: pkgs/applications/office/paperless-ngx/default.nix [hunk 3]
```diff
@@ -46,14 +54,26 @@ let
...
+    buildInputs = [
+      cairo
+      giflib
+      libjpeg
+      libpng
+      librsvg
+      pango
+      pkg-config
+    ] ++ lib.optionals stdenv.isDarwin [
+      mupdf
```
Comment (SuperSandro2000) [inline pkgs/applications/office/paperless-ngx/default.nix:74]: Not sure where this is coming from

File: pkgs/applications/office/paperless-ngx/default.nix [hunk 4]
```diff
@@ -46,14 +54,26 @@ let
...
       python3
     ] ++ lib.optionals stdenv.isDarwin [
       xcbuild
     ];
 
+    buildInputs = [
+      cairo
+      giflib
+      libjpeg
+      libpng
```
Comment (SuperSandro2000) [inline pkgs/applications/office/paperless-ngx/default.nix:69]: According to https://github.com/Automattic/node-canvas#compiling we only need this on darwin/bsd.

**Review by SuperSandro2000 — COMMENTED** [review]
I tried replicating your changes and generally I agree with all of them.

**Comment by SuperSandro2000** [comment]
Also can we do something about the log spam in paperless-ngx-frontend

File: pkgs/applications/office/paperless-ngx/default.nix (hunk 3 above)
Comment (ajaxbits) [reply pkgs/applications/office/paperless-ngx/default.nix:74]: Had some errors around `mutool` on aarch64-Darwin. But I'll reevaluate to see if I need this after putting pkg-config in the right place.

File: pkgs/applications/office/paperless-ngx/default.nix (hunk 2 above)
Comment (ajaxbits) [reply pkgs/applications/office/paperless-ngx/default.nix]: Confirmed!

File: pkgs/applications/office/paperless-ngx/default.nix [hunk 5]
```diff
@@ -46,14 +54,26 @@ let
...
+      librsvg
+      pango
+      pkg-config
```
[the remark this answers is not in the document]
Comment (ajaxbits) [reply pkgs/applications/office/paperless-ngx/default.nix]: addressed

File: pkgs/applications/office/paperless-ngx/default.nix [hunk 6]
```diff
@@ -46,14 +54,26 @@ let
...
+    ] ++ lib.optionals stdenv.isDarwin [
+      mupdf
+    ];
+
     postPatch = ''
       cd src-ui
     '';
```
[the remark this answers is not in the document]
Comment (ajaxbits) [reply pkgs/applications/office/paperless-ngx/default.nix]: addressed

File: pkgs/applications/office/paperless-ngx/default.nix (hunk 3 above)
Comment (ajaxbits) [reply pkgs/applications/office/paperless-ngx/default.nix]: Looks like this is actually an error with the `img2pdf` python package building on aarch64-darwin...

Here's what I get on my machine:

```
nixpkgs on paperless-ngx-210-bump [$] took 2m39s
❯ nix build path:(pwd)#python311Packages.img2pdf
error: builder for '/nix/store/9zks1lvjp7vb573bc7977x1dfhwnpk0i-python3.11-img2pdf-0.5.1.drv' failed with exit code 2;
       last 10 log lines:
       > /nix/store/ai4hd8f1xhr0rfjdr17bxx1rwi42sx97-python3-3.11.6/lib/python3.11/subprocess.py:466: in check_output
       >     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
       > /nix/store/ai4hd8f1xhr0rfjdr17bxx1rwi42sx97-python3-3.11.6/lib/python3.11/subprocess.py:571: in run
       >     raise CalledProcessError(retcode, process.args,
       > E   subprocess.CalledProcessError: Command '['mutool', '-v']' died with <Signals.SIGABRT: 6>.
       > =========================== short test summary info ============================
       > ERROR src/img2pdf_test.py - subprocess.CalledProcessError: Command '['mutool', '-v']' died with <Signal...
       > !!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
       > =============================== 1 error in 5.12s ===============================
       > /nix/store/2a1pjx5crgygz7wbvl1lmp49n6apfp8r-stdenv-darwin/setup: line 1568: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/9zks1lvjp7vb573bc7977x1dfhwnpk0i-python3.11-img2pdf-0.5.1.drv'.
```

File: pkgs/applications/office/paperless-ngx/default.nix (hunk 4 above)
Comment (ajaxbits) [reply pkgs/applications/office/paperless-ngx/default.nix]: Confirmed that I was able to get rid of most of these unnecessary packages. However, we *do* need `pango` as a buildInput in order to get the package to build. Putting it in nativeBuildInputs didn't work for me.

**Comment by ajaxbits** [comment]
Result of `nixpkgs-review pr 272563` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>paperless-ngx</li>
  </ul>
</details>

File: pkgs/applications/office/paperless-ngx/default.nix (hunk 3 above)
Comment (ajaxbits) [reply pkgs/applications/office/paperless-ngx/default.nix]: And looks like it's actually a `mupdf` error:

```
nixpkgs on  paperless-ngx-210-bump [$] took 2m12s
❮ nix build path:(pwd)#mupdf

nixpkgs on  paperless-ngx-210-bump [$] took 2m12s
❯ result-bin/bin/mutool -v
dyld[85912]: Library not loaded: build/shared-release/libmupdf.dylib
  Referenced from: <no uuid> /nix/store/53anxhdg9mqh8cj568hrqd8w8yfxp07f-mupdf-1.23.5-bin/bin/mutool
  Reason: tried: 'build/shared-release/libmupdf.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSbuild/shared-release/libmupdf.dylib' (no such file), 'build/shared-release/libmupdf.dylib' (no such file), '/usr/local/lib/libmupdf.dylib' (no such file), '/usr/lib/libmupdf.dylib' (no such file, not in dyld cache)
fish: Job 1, 'result-bin/bin/mutool -v' terminated by signal SIGABRT (Abort)
```

Comment (ajaxbits) [reply pkgs/applications/office/paperless-ngx/default.nix]: I'm not sure I have the firepower to debug this issue in a timely manner. Should it be a blocker for merging this PR?

**Comment by ajaxbits** [comment]
Ok, I think this is ready for review as it builds nicely on x86_64 linux and tests all pass. There are still compilation errors on aarch64-darwin, but I don't know if I can track down the error in a timely manner. Details [here](https://github.com/NixOS/nixpkgs/pull/272563#discussion_r1419750168). 

Let me know if the above issues should block this PR. But in the meantime, I'm marking this

File: pkgs/applications/office/paperless-ngx/default.nix (hunk 3 above)
Comment (ajaxbits) [reply pkgs/applications/office/paperless-ngx/default.nix]: Opened issue #272789 for this.

Comment (SuperSandro2000) [reply pkgs/applications/office/paperless-ngx/default.nix]: If the relevant package was broken like this before, we can skip this for now.

**Review by SuperSandro2000 — APPROVED** [review]

**Review by erikarvstedt — CHANGES_REQUESTED** [review]
~~Please don't move `postPatch`. It's idiomatic to place this next to other phase related definitions, as before.~~
Corrected by SuperSandro, below.

**Review by erikarvstedt — CHANGES_REQUESTED** [review]
[v2.1.1](https://github.com/paperless-ngx/paperless-ngx/releases/tag/v2.1.1) has been released.

**Comment by SuperSandro2000** [comment]
> Please don't move `postPatch`. It's idiomatic to place this next to other phase related definitions, as before.

Nope, the standard place for it is right after src and before any hashes it influences.

File: pkgs/applications/office/paperless-ngx/default.nix (hunk 3 above)
Comment (ajaxbits) [reply pkgs/applications/office/paperless-ngx/default.nix]: Looks like this will also be resolved with this pr: https://github.com/NixOS/nixpkgs/pull/272807

**Comment by ajaxbits** [comment]
Result of `nixpkgs-review pr 272563` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>paperless-ngx</li>
  </ul>
</details>

**Comment by ajaxbits** [comment]
@erikarvstedt, just updated the code for v2.1.1, and has been built and tested successfully!

**Review by leona-ya — APPROVED** [review]

**Review by leona-ya — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]
